src/TestFramework/Coverage/CoveredFileDataFactory.php

```diff
@@ -0,0 +1,128 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\TestFramework\Coverage;
+
+use Infection\Mutation\MutationGenerator;
+use Infection\TestFramework\Coverage\JUnit\TestFileDataAdder;
+use Infection\TestFramework\Coverage\XmlReport\PhpUnitXmlCoverageFactory;
+use Symfony\Component\Finder\SplFileInfo;
+use Webmozart\Assert\Assert;
+
+/**
+ * Assembles a ready feed of CoveredFileData from different sources. Feeds data into MutationGenerator. Does not known about differences between adapters and what not.
+ *
+ * @see MutationGenerator
+ * @see PhpUnitXmlCoverageFactory
+ *
+ * @internal
+ */
+final class CoveredFileDataFactory implements CoveredFileDataProvider
+{
+    /** @var CoveredFileDataProvider|PhpUnitXmlCoverageFactory */
+    private $primaryCoverageProvider;
+
+    /** @var iterable<SplFileInfo> */
+    private $sourceFiles;
+
+    private $filter;
+
+    private $testFileDataAdder;
+
+    /**
+     * @param iterable<SplFileInfo> $sourceFiles
+     */
+    public function __construct(
+        CoveredFileDataProvider $primaryCoverageProvider,
+        TestFileDataAdder $testFileDataAdder,
+        CoveredFileNameFilter $filter,
+        iterable $sourceFiles
+    ) {
+        $this->primaryCoverageProvider = $primaryCoverageProvider;
+        $this->testFileDataAdder = $testFileDataAdder;
+        $this->filter = $filter;
+        $this->sourceFiles = $sourceFiles;
+    }
+
+    /**
+     * @return iterable<CoveredFileData>
+     */
+    public function createCoverage(): iterable
+    {
+        $coverageFeed = $this->primaryCoverageProvider->createCoverage();
+
+        $filteredCoverageFeed = $this->filter->filter($coverageFeed);
+
+        $readyCoverageFeed = $this->testFileDataAdder->addTestExecutionInfo($filteredCoverageFeed);
+
+        if (is_array($this->sourceFiles) && $this->sourceFiles === []) {
+            // The case where only covered files are considered
+            return $readyCoverageFeed;
+        }
+
+        return $this->appendUncoveredFiles($readyCoverageFeed);
+    }
+
+    /**
+     * @param iterable<CoveredFileData> $coverage
+     *
+     * @return iterable<CoveredFileData>
+     */
+    private function appendUncoveredFiles(iterable $coverage): iterable
+    {
+        $seenFiles = [];
```

Comment:
`$checkedFiles`?

Reply:
Well, we don't do any checking with them.

Reply:
`$filesSeen`?

Reply:
Done